### ISSUE TITLE
Youtube iframe and lazyload with lazysizes

### ISSUE THREAD

**Issue by Lofesa** [Really-Simple-Plugins/complianz-gdpr#339]
Hi
We use the 6.0.1 version of this pligin and a plugin based on aFarcas lbrary for lazyload images and iframes (`https://github.com/aFarkas/lazysizes`).
This libray put a class "lazyload" to the iframe and changes src with data-src for the video url.
When cookies are not acepted all works as spected, image placeholder and button to accept marketing cookies is present, but when cookies are accepted no video is showed, only the image placeholder, and js throw a error in complianz.min.js file.
If lazyload for iframes is disabled all works, but if cookies are acepted the iframe resources are loaded at page start.
As a workaround I have done some changes to plugin files and it works now, when cookies are not accepted the iframe don´t load and when accepted the iframe is lazyloaded.
The changes I have done are:
comment the line 377 in the file `class-cookie-blocker.php`
change src to data-src in lines 543 and 561 in the file `complianz.js` 

Can these changes be included in the plugin in some way?
Some like:
In php file:

`$aFarcas = str_contains($new,'lazyload');
.
.
if (!$aFarcas) $new = $this->replace_src( $new, apply_filters( 'cmplz_source_placeholder', 'about:blank' ) );`

In js file:

`let origin ='src';
if (obj.classlist.contains('lazyload')) { let origin = 'data-src';}
obj.setAttribute('origin', src);`

Thanks in advance

**Comment by rlankhorst** [comment]
Please note that without the lazy loading, the plugin will still act as lazy loader, because the iframes are blocked by default, then unlocked after the page has loaded completely. 

If you want the lazy loading feature, because it will then also delay for example until the content is scrolled into view: is it possible to change the class to a data attribute 

loading="lazy"?

This is the browser native lazy loading attribute, and is already built in.

**Comment by Lofesa** [comment]
Thx for your response.

And yes, blocking the iframe until cookies are accepted acts as a lazyload stuff, but I´m talking about a user with cookies accepet, when this user come to the page with youtube iframe, all youtube stuff is loaded because nothing is blocking the iframe..

So called native lazyloading is a ... It takes the size of the page to be loaded or delayed whit the "detected" speed and set a very large size even in poor connections making it unusefull.

Plus, Firefox don´t support native lazyload for iframes, Safari need to be enabled by the user, Safari on iOS don´t support it.... so I think make sense support some of the most widespread lazyload plugins, and much them uses the aFarcas library.

**Comment by rlankhorst** [comment]
@Lofesa even with cookies accepted, the content is first blocked, then after page load the content is unlocked. so the cookie blocker always works as a lazy load feature, even if the cookies are accepted.

**Comment by Lofesa** [comment]
@rlankhorst  Ummm no. When you visit a page whit iframe not lazyloaded all the stuff from youtube is downloaded (and cookies accepted of course)
You can try it here:

iframe not lazyloaded:

`https://intersindicalrm.org/65-medidas-para-el-desarrollo-economico-y-social-de-la-region/`

this load all the youtube stuuf whe you go with cookies acepted

iframe lazyloaded:

`https://intersindicalrm.org/priorizar-la-consolidacion-acabar-con-la-temporalidad/`

this other don´t load nothing from youtube until scroll down.

The placeholder image because is not in a img element (or backgruond css) is not lazyloaded.

**Comment by rlankhorst** [comment]
It's not lazy loaded in the sense that it waits until scrolled, what I meant is that the page is loaded in a blocked state, then, as the page is loaded completely, after page load has completed, the complianz script unlocks the content. Which is a form of lazy loading. 

But I understand you want it to load only after a user scrolls. 

As this isn't a generic issue, I would prefer it to be included in an integration. For this to work it needs to be a WordPress plugin. Is that the case? In that case, let me know which one it is, and we can create an integration like described here:

https://complianz.io/developers-guide-for-third-party-integrations/

Alternatively, a mu-plugin is an option

**Comment by Lofesa** [comment]
Hi
The plugin is Lazy Loader (https://es.wordpress.org/plugins/lazy-loading-responsive-images/).
But I think the way of integration is not a valid one because the files I have changed are "core" files from Coplianz plugin.
You have a more coomprensive view of the whole picture.....

**Comment by rlankhorst** [comment]
@Lofesa thanks, we'll build an integration for this plugin, I'll get back to you on this.

**Comment by rlankhorst** [comment]
@Lofesa I've been running some tests with the lazy load plugin (https://wordpress.org/plugins/lazy-loading-responsive-images/). 

I've enabled lazy loading for iframe and videos, and enabled the Youtube integration in Complianz. But I couldn't reproduce the issue you described: after accepting cookies the video shows, no js error is thrown. 

Are you using other settings perhaps?

**Comment by Lofesa** [comment]
Hi
Sounds stange... the plugin must add a class (lazyload) and change src to data-src, so when cookiebanner/js/complianz.min.js seek for a src element (line 543 of the unminyfied file to test if autoplay is set) there is nothing and throw an error (in line 1777 of the unminifyed file)  and stop javascript execution.

Can you share the url where the test is done?

**Comment by rlankhorst** [comment]
I've set up a sandbox with complianz and the lazy loader plugin:
https://s-qcxnz77fxykkl.eu1.wpsandbox.org/about/

**Comment by Lofesa** [comment]
Hi
An access code is needed

**Comment by rlankhorst** [comment]
@Lofesa sorry, forgot about that. 165

**Comment by Lofesa** [comment]
Hi again @rlankhorst 

I can see whats happen now.

The video is near to the viewport. The plugin has a 500px area near the viewport that images or iframes are loaded so no data-src here aka no error.

In my site have a 200px area by custom parameter.

Put comtent before the video and test that class in the iframe still "lazyload" if you see "lazyloaded" add more content above the video.
Test "lazyload" class has not changed to "lazyloaded" without scrolling and the data-src is here,  then the error comes.
I have manually changed src to data-src and the error is here. (see the image)

To set a short area size, in wordpress admin -> settings -> media, at the bottom of the page you have a text area:  Modify the default config.
Add:

window.lazySizesConfig = window.lazySizesConfig || {};
lazySizesConfig.expand = 200;

or whatever number you will try, but the objective is getting the "lazyload" class without scrolling and the data.-src not changed to src.


![Captura de pantalla 2022-01-27 191103](https://user-images.githubusercontent.com/22342793/151419610-3f5c40a1-9f07-46a3-b672-2e2edb73d192.jpg)

**Comment by Lofesa** [comment]
Here can you test it:
https://u8903-systematization-660.sites.qsandbox.com/
code(123)

**Comment by rlankhorst** [comment]
I haven't yet been able to reproduce the error you have there. I've now added more content, and added the config you posted:
https://s-qcxnz77fxykkl.eu1.wpsandbox.org/about/

On load (consent given) the class is 'lazyload', I see a data- src. After scrolling, the class changes to lazyloaded. The video shows normally. 

I chose both lazy load for iframes and videos, don't know if that makes a difference? Is there anything else I need to change to reproduce the issue?

**Comment by Lofesa** [comment]
Hi
I don´t know why, but in your test the src still here, not changed to data-src.
My choices for the lazyload plugin are:
Enable lazy loading for iframes
Include lazysizes unveilhooks plugin
Enable lazy loading for inline background images
Enable option to disable plugin per page/post
Process the complete markup
Modify the default config = window.lazySizesConfig = window.lazySizesConfig || {};
lazySizesConfig.expand = 200;

Anyway, you have the issue reproduced in:
https://u8903-systematization-660.sites.qsandbox.com/
The code for access is 123

My test:

![Captura de pantalla 2022-01-29 102155](https://user-images.githubusercontent.com/22342793/151655524-aba75c85-9e84-48ea-9a0b-5c6bd13541c5.jpg)

Yours:


![Captura de pantalla 2022-01-29 102231](https://user-images.githubusercontent.com/22342793/151655541-56bf5881-51bb-4588-ac9d-c9acabc99841.jpg)

**Comment by rlankhorst** [comment]
@Lofesa 
As I can't create a fix if I can't reproduce it, I need to be able to reproduce the issue exactly. I've succeeded in that: 
the cause of the issue appears to be "Process the complete markup". 

The issue was caused by the 'about:blank' of the iframe src. There is a function which retrieves the url paramaters of the video (like autoplay etc). In this case, this function failed because the url is empty. I've created a fix that catches the existence of an empty url. With that fix, the video loads without issues for me. Can you check if this resolves the issue for you as well?

https://github.com/Really-Simple-Plugins/complianz-gdpr/tree/catch-empty-url-in-cmplz_get_url_parameter

**Comment by Lofesa** [comment]
Hi
The only files you have changed are the cookiebanner/js7complianz.j and the minifyes version.
I have copied the minified version and changed it in my site, the a new issue comes here:

The button for blocking text can´t be created and fails even cookies accepted.

![Captura de pantalla 2022-01-31 135736](https://user-images.githubusercontent.com/22342793/151797592-88aadcf9-8bfb-4d1a-83d0-e4f6ee50602b.jpg)

![Captura de pantalla 2022-01-31 135803](https://user-images.githubusercontent.com/22342793/151797607-63352d59-7474-4f78-89d7-90ea0455a108.jpg)

**Comment by rlankhorst** [comment]
@Lofesa This change has been branched of the 6.1 version, which includes several other changes (see changelog), Please upgrade the entire plugin, otherwise this can result in issues.

**Comment by rlankhorst** [comment]
@Lofesa the error shown in your console is not related to the data-src, but means the html for the blocked content container is not present. It's the same error as in Firefox.  I saw this error once, but after deactivating and re-activating the "lazy load background images" option, the error is gone. Possibly this feature is not compatible with how complianz works. 

As the placeholder works with background images as well, it would seem tricky to use lazy load on background images. 

 A page refresh should not be necessary, as the src is dynamically replaced and loaded after consent. The cookies merely make sure the behaviour is persistent. 

Can you try deactivating the lazy load for background images feature?

**Comment by Lofesa** [comment]
Hi
Even with background images lazyload deactivated the error come here.

**Comment by rlankhorst** [comment]
Hi @Lofesa, I have re-created a sandbox, with your settings, and the latest branch. Even with the lazy loading background images this seems to work now:

https://s-q3ant5ks7c57.eu1.wpsandbox.org/youtube/
access 437

**Comment by Lofesa** [comment]
Hi
The error is here

![Captura de pantalla 2022-02-01 185337](https://user-images.githubusercontent.com/22342793/152023732-1e25b6d1-9c21-4a1a-a7f8-3f276aeb2400.jpg)

**Comment by rlankhorst** [comment]
the error seems only to occur on a new page load, not when the blocked video is consented. Then it loads fine. I have to check why there is a difference between the two methods. Possibly it is a race condition, where a script is fired too early or too late.

**Comment by rlankhorst** [comment]
Ok, based on the race condition, I now understand what was happening. The lazy loader sometimes tries to change the url after complianz. I have now created an integration for lazy loader:

https://github.com/Really-Simple-Plugins/complianz-gdpr/compare/catch-empty-url-in-cmplz_get_url_parameter?expand=1

it works for me on reload now on the sandbox as well. Let me know!

**Comment by Lofesa** [comment]
Hi
The error still here, the same error. But....

The error comes when at start of the page, then scroll down until the video comes to the viewport, then the error fires (1st image)
Whit the video in the viewport do a reload and then no error here, only the button complianz set for blocked videos appear and disappears brieftly (2nd image)

In both cases the video is not lazyloaded, all the youtube stuff is downloaded at start.

![Captura de pantalla 2022-02-02 220254](https://user-images.githubusercontent.com/22342793/152237155-d5b49592-f76e-41bd-8ea3-adda1f5d365c.jpg)


![Captura de pantalla 2022-02-02 220414](https://user-images.githubusercontent.com/22342793/152237183-6778ca08-e2e7-444f-97fb-ee03e06c2ede.jpg)

**Comment by rlankhorst** [comment]
Are you sure it isn't caching? Did you try an anonymous/private window?

**Comment by Lofesa** [comment]
Yes, I use a chrome that deletes all on exit.

**Comment by rlankhorst** [comment]
I have created a completely new sandbox:
https://mawehe.us2.instawp.xyz/index.php/youtube/

I can't see any issues. I've tried both Chrome and Safari, and accepting through the banner, reload after accept etc. Do you also have the issue on the new sandbox?

**Comment by Lofesa** [comment]
In incognito mode, when cookies are acepted, I don´t know why the:
```
<div class="wp-block-embed__wrapper" data-placeholder_class_index="1">
<noscript>
```
don´t have these class `cmplz-placeholder-1 cmplz-blocked-content-container` and throw the error when scroll to the video

![Captura de pantalla 2022-02-03 111511](https://user-images.githubusercontent.com/22342793/152323784-6711b9b7-75c0-4d4f-8400-264a5175064c.jpg)

Anyway, the complianz plugin still set the src=youtube-url attribute, this makes all youtube stuff download at start of the page and makes lazyload unusable.
I think the complianz.min.js must check there is a data-src and if present, don´t put the src attribute and let the lazyload take care of it.

**Comment by Lofesa** [comment]
Let´s make a recap on this. despite the js error throwed by no class preesent or for seaching for attributtes that are not here:

1.- We have an iframe element:
`<iframe..... src=youtube-url.......></iframe>`

2.- Then Complianz changes some things:
`<iframe..... src=about:blanck data-src-cmplz=youtube-url.......></iframe>`

3.- Now the lazyload plugin makes news changes:
`<iframe..... data-src=about:blank data-src-cmplz=youtube-url.......></iframe>`

Until here is perfect to block youtube to donwload all their stuff and set any cookie. 
Things go wrong when cookies are accepted and both plugins make news changes:

4.- Complianz do their work by javascript:
`<iframe..... data-src=about:blank data-src-cmplz=youtube-url src=youtube-url.......></iframe>`
This makes youtube download all their stuff and makes lazyload unusable.

5.-Lazyload plugin does their work:
`<iframe..... data-src=about:blank data-src-cmplz=youtube-url src=youtube-url.......></iframe>`
Unveils the youtube player with the url set in data-src, so a blank space because data-src is about:blank.

What we need here?
Steps 1-3 are ok. Step 4 is the key. If cookies are accepted, we NOT need a src=youtube-url attribute because this break the lazyload function, we need restore the data-src=youtube-uel attribute.

All this is WITHOUT the changes you have made to the javascript file.

What can be a solution?

```
} else if (tagName==='IFRAME'){
			obj.classList.add('cmplz-activated' );
                       ** let label = obj.hasAttribute('data-src');**
                        **let source = 'src';**
                        **if(label === true){ source='data-src'};**
			let src = obj.getAttribute('data-src-cmplz');
			//check if there's an autoplay value we need to pass on
			**let autoplay = cmplz_get_url_parameter(obj.getAttribute(source), 'autoplay');**
			if ( autoplay === '1' ) src = src + '&autoplay=1';

			obj.addEventListener('load', (event) => {
				//handle browser native lazy load feature
				if ( obj.getAttribute('data-deferlazy') ) {
					obj.setAttribute('loading', 'lazy');
				}

				//we get the closest, not the parent, because a script could have inserted a div in the meantime.
				let blocked_content_container = obj.closest('.cmplz-blocked-content-container');
				let cssIndex = blocked_content_container.getAttribute('data-placeholder_class_index');
				blocked_content_container.classList.remove('cmplz-blocked-content-container');
				blocked_content_container.classList.remove('cmplz-placeholder-' + cssIndex);
				obj.classList.remove('cmplz-iframe-styles');
				obj.classList.remove('cmplz-iframe');
				obj.classList.remove('video-wrap');
			});
			**obj.setAttribute(source, src);**

```
Note: Between `**text**` are lines addend or changed.

Note 1: ` let label = obj.hasAttribute('data-src');`  maybe this can be change to test if iframe has an attribute in an array and set the varible source accordlyng for some others plugins like WPRockect, I thik it uses data-lacy-src attribute., and then restore their original value, the youtube-url. I don´t know how to do this as far as I don´t be a js coder

**Comment by rlankhorst** [comment]
checking for a generic name like data-src or a lazyload class is too risky, the risk of conflicts with other plugins when using this on 300000 websites is too high. I think I have a solution: I have added a function in the integration that adds the data-src attribute with a filter. 

When checking the source for https://mawehe.us2.instawp.xyz/index.php/youtube/ I see no src="" until the video is scrolled into view.

**Comment by Lofesa** [comment]
When cookies are acepted, there is a src=youtube url, see the image.
1 red circle the noscript element with src=abount:blank set by Complinaz before the lazyload plugin manipulates the iframe
2 red circle the src=youtube-url set by Complianz at page start

![Captura de pantalla 2022-02-03 130958](https://user-images.githubusercontent.com/22342793/152340627-569a00f0-ce7e-462b-9d8e-0e1ca73b24df.jpg)

In the network tab all the stuff from youtube is here.... because the src=youtube-url is here.
I can see you use the native lazyload..... thats is unusefull. Native lazyload set the viewport in function of "detected" network speed, the minimun size added to the viewport in a fast network is 2000px and in a slow 2G 6000px. You can view it in the Chromium code here

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/frame/settings.json5;drc=e8f3cf0bbe085fee0d1b468e84395aad3ebb2cad;l=971-1003

I share yorurs concerns about compat with the whole universe of plugins, and maybe a filter can be used, but not server side, because there are other whole universe of page cache plugins that serve the html in the cache without firing wordpress stuff.
The filter must be run client side aka javascript at the moment.

**Comment by rlankhorst** [comment]
In the current implementation I add a data-target attribute (will be prefixed to prevent conflicts) which allows the javascript to recognise lazy loader. 

It seems to work now, only if you scroll down, then consent, it doesn't work yet. I'll look at that tomorrow.

**Comment by rlankhorst** [comment]
The issue I run into with this, is that if I use your method:

- not replacing src with "about:blank"
- replacing data-src with data-src-cmplz

As soon as the the lazy load plugin is triggered, the video is loaded, before consent, because the data-src is available for the lazy load plugin. It just loads the video, and the placeholder is on top of it so it looks blocked. But it isn't. So that is not enough. 

So, to ensure the video is blocked until consent, the src has to be replaced to about:blank. But if I do this, the lazy load plugin can't handle the load. 

I have now added a javascript hook which replaces the 'lazyloaded' class with 'lazyload' after consent category marketing fires. This ensures that the lazy load plugin knows it still has work to do.  

You can view the result here:
https://mawehe.us2.instawp.xyz/index.php/youtube/

https://github.com/Really-Simple-Plugins/complianz-gdpr/tree/catch-empty-url-in-cmplz_get_url_parameter

**Comment by Lofesa** [comment]
Hi
I think is cause you have loading="lazy" attribute, then native lazyloading works before the plugin.
Can you try to not set the native lazyload attribute?

P.D. I feel all this.
I'm sorry to be taking up your time and bothering you with this matter.
But I think that Complianz is a good plugin doing its job and I would like it to be even better

**Comment by rlankhorst** [comment]
I have disabled the option "Include lazysizes native loading plugin" in Lazy Loader now. 

No problem at all, we'll get there in the end :)

**Comment by Lofesa** [comment]
Now it works .
A stetic question: Why the button element (these used when the cookies are not accepted in the placeholder image) is here when cookies are accepted?
In the image cookies are accepted, the button don´t show at the screen, but the dom element is here.

![Captura de pantalla 2022-02-04 145909](https://user-images.githubusercontent.com/22342793/152541449-0419ad17-de81-4d08-87cf-9029aa4b7547.jpg)

**Comment by rlankhorst** [comment]
The html is loaded blocked by default, the javascript inserts the placeholder html for each blocked element. On load, the elements are unlocked. So the button is loaded by default, then hidden with css if consent is given.
